File: LeetCode/ByteDance/BinSearch/MySqrt.cpp
class Solution {
  public:
    int mySqrt(int x) {
        if (x < 2)
            return x;
        int l = 2, r = x / 2;
        while (l <= r) {
            long m = (l + r) >> 1;
            long t = m * m;
            if (t > x)
                r = m - 1;
            else if (t < x)
                l = m + 1;
            else
                return m;
        }
        return r;
    }
};

int main() {}

File: LeetCode/Explore/BinSearch/mySqrt.go
package leetcode

func mySqrt(x int) int {
	if x < 2 {
		return x
	}
	var l, r = 2, x >> 1
	for l <= r {
		var m = l + (r-l)>>1
		var t = m * m
		if t == x {
			return m
		} else if t < x {
			l = m + 1
		} else {
			r = m - 1
		}
	}
	return r
}
